api/src/main/java/io/opentelemetry/trace/TraceState.java

```diff
@@ -235,26 +235,53 @@ public static Entry create(String key, String value) {
   // Key is opaque string up to 256 characters printable. It MUST begin with a lowercase letter, and
   // can only contain lowercase letters a-z, digits 0-9, underscores _, dashes -, asterisks *, and
   // forward slashes /.  For multi-tenant vendor scenarios, an at sign (@) can be used to prefix the
-  // vendor name.
+  // vendor name. The tenant id (before the '@') is limited to 240 characters and the vendor id is
+  // limited to 13 characters. If in the multi-tenant vendor format, then the first character
+  // may additionally be digit.
+  //
   // todo: benchmark this implementation
   private static boolean validateKey(String key) {
-    if (key.length() > KEY_MAX_SIZE || key.isEmpty() || !isNumberOrDigit(key.charAt(0))) {
+    if (key.length() > KEY_MAX_SIZE || key.isEmpty() || !isLowercaseLetterOrDigit(key.charAt(0))) {
       return false;
     }
     int atSeenCount = 0;
     for (int i = 1; i < key.length(); i++) {
       char c = key.charAt(i);
-      if (!isNumberOrDigit(c) && c != '_' && c != '-' && c != '@' && c != '*' && c != '/') {
+      if (!isLowercaseLetterOrDigit(c)
+          && c != '_'
+          && c != '-'
+          && c != '@'
+          && c != '*'
+          && c != '/') {
         return false;
       }
-      if ((c == '@') && (++atSeenCount > 1)) {
-        return false;
+      int atPosition = 0;
+      if (c == '@') {
+        atPosition = i;
+        // tenant id must be 240 characters or less
+        if (i > 240) {
+          return false;
+        }
+        atSeenCount++;
+        if (atSeenCount > 1) {
+          return false;
+        }
       }
+      if (atSeenCount == 1) {
+        // vendor id must be 13 characters or less
+        if ((i - atPosition) > 13) {
```

Comment:
I may just be misgrokking, but isn't `atPosition` `0` for most of the loop rather than the actual `atPosition`? If my reading is right, my above suggestion would probably solve it but it also means there's a unit test missing for the current buggy state.

Reply:
`atSeenCount` is incremented only when `atPosition` changes to `i`. This way, this if is checked only for `atPosition` > 0.

Reply:
Yeah but `atPosition` is declared inside the loop, which I think is the original bug. So except for when key[i] == '@', `atPosition == 0` for the remainder of the loop since it's initialized to `0`. Sorry if I'm missing it though I've tried to stare long and this is what I came up with.

Reply:
Oh! You are totally right! I miss that bit! We are indeed missing a test here!

Reply:
I'm not following the problem. Can you give me a key that we're not handling correctly, so I can add a unit test for the issue?

Reply:
Note, I'm rewriting this chunk of code to make it clearer. But, having an example key that we're not handling correctly would be very useful (because I don't see the issue).

Reply:
I've pushed the rewrite. Let me know if that helps at all with understanding what's going on in here. 

And, please give me a concrete key example case we're not handling correctly if there is one.  :)

Reply:
I'm pretty sure `12345678901234567890@nr` is valid but fails when I added this to your branch. Basically any time the tenant id is longer than 13 characters, since the current logic is resetting the atPosition to the beinning of the string.

My [suggestion](https://github.com/open-telemetry/opentelemetry-java/pull/1561#discussion_r473635930) of not bothering with `atPosition` still stands :) 

```


  @Test
  void testValidLongVendorKey() {
    TraceState result = EMPTY.toBuilder().set("12345678901234567890@nr", FIRST_VALUE).build();
    assertThat(result.get("12345678901234567890@nr")).isEqualTo(FIRST_VALUE);
  }
```